### PR TITLE
State blocks sideband upgrade adding is_send/receive/epoch

### PR DESCRIPTION
**Pull request by guilhermelawless**
This changes how the epoch byte in the state block sideband is [de]serialized, to also store these flags in the 3 most significant bits: `[is_send | is_receive | is_epoch | epoch (5 bits)]`

The upgrade is done in-place but takes a long time due to having to a random read to retrieve the previous block. A vacuum is performed at the end.

The upgrade does the following:
- Iterate through the state_blocks table
- For each block grab the previous block to determine the subtype of the current block
- Serialize the new sideband
- `mdb_cursor_put` with `MDB_CURRENT`

I was expecting no size growth due to using [MDB_CURRENT](http://www.lmdb.tech/doc/group__mdb__put.html#ga92f7832a496213db0f03105e6fa9afae), but no such luck. [mdb_cursor_put](http://www.lmdb.tech/doc/group__mdb.html#ga1f83ccb40011837ff37cc32be01ad91e) docs say:
> MDB_CURRENT - replace the item at the current cursor position. The key parameter must still be provided, and must match it. If using sorted duplicates (MDB_DUPSORT) the data item must still sort into the same place. This is intended to be used when the new data is the same size as the old. Otherwise it will simply perform a delete of the old record followed by an insert.

Since the new sideband deserialization discards the MSBs when reading the epoch, the upgrade can safely be stopped and restarted (from the beginning), and no additional versioning was necessary. This behavior is tested in the upgrade test.

However, it was necessary to move up the `block_sideband` class definition in order to have `block_w_sideband` declared in the same file (declaring in `versioning.hpp` would require saving a copy of the `block_sideband` class in `versioning.cpp`).

Also added a validation step of these new fields in the `--debug_validate_blocks` CLI.

-----

**Best case scenario test** (on a SATA III SSD)
Tested with @SergiySW 's daily ledger (which is rebuilt with https://github.com/nanocurrency/nano-node/pull/2435):
- Initial size 17GB
- Initial v16->v17 upgrade takes 2 minutes, basically no growth
- v17->v18 upgrade takes 45 minutes, growing to 29GB
- Post-upgrade vacuum brings it back to 17GB

This means the total space necessary for the best-case scenario was 29+17 = 46GB.

### CODE REVIEW

**Comment by guilhermelawless** [comment]
Log example:

```
[2020-Jan-31 12:51:39.787604]: Preparing v16 to v17 database upgrade...
[2020-Jan-31 12:52:05.966315]: Confirmation height frontier set for 200k accounts
[2020-Jan-31 12:52:27.812587]: Confirmation height frontier set for 400k accounts
[2020-Jan-31 12:52:48.697944]: Confirmation height frontier set for 600k accounts
[2020-Jan-31 12:53:09.644635]: Confirmation height frontier set for 800k accounts
[2020-Jan-31 12:53:29.040246]: Confirmation height frontier set for 1000k accounts
[2020-Jan-31 12:53:43.478123]: Finished upgrading confirmation height frontiers
[2020-Jan-31 12:53:43.480816]: Preparing v17 to v18 database upgrade...
[2020-Jan-31 12:55:01.512548]: Database sideband upgrade 1 million state blocks upgraded (out of 39024411)
[2020-Jan-31 12:56:13.118961]: Database sideband upgrade 2 million state blocks upgraded (out of 39024411)
[2020-Jan-31 12:57:27.724141]: Database sideband upgrade 3 million state blocks upgraded (out of 39024411)
(...)
[2020-Jan-31 13:38:40.215407]: Database sideband upgrade 39 million state blocks upgraded (out of 39024411)
[2020-Jan-31 13:38:41.990507]: Finished upgrading the sideband
```

File: nano/secure/blockstore.cpp [hunk 1]
```diff
@@ -87,7 +128,7 @@ bool nano::block_sideband::deserialize (nano::stream & stream_a)
 		boost::endian::big_to_native_inplace (timestamp);
 		if (type == nano::block_type::state)
 		{
-			nano::read (stream_a, epoch);
+			details.deserialize (stream_a);
```
Comment (wezrule) [inline nano/secure/blockstore.cpp]: You should check the return argument here in case of failure. Or in `block_details::deserialize` don't use a try-catch and let the exception propagate (with an appropriate comment)

File: nano/secure/blockstore.hpp [hunk 2]
```diff
@@ -16,6 +16,62 @@
...
+	static constexpr size_t size ()
+	{
+		return 1;
+	}
+	bool operator== (block_details const & other_a) const
+	{
+		return epoch == other_a.epoch && is_send == other_a.is_send && is_receive == other_a.is_receive && is_epoch == other_a.is_epoch;
+	}
+	void serialize (nano::stream &) const;
+	bool deserialize (nano::stream &);
```
Comment (wezrule) [inline nano/secure/blockstore.hpp]: Some functions are defined inline in the class definition but `serialize`/`deserialize` aren't. Would be good to move them all out, if only for consistency.

File: nano/secure/blockstore.hpp [hunk 3]
```diff
@@ -279,6 +335,21 @@ class db_val
...
+		nano::state_block_w_sideband state_block_w_sideband;
+		state_block_w_sideband.state_block = std::make_shared<nano::state_block> (error, stream);
+		assert (!error);
+
+		state_block_w_sideband.sideband.type = nano::block_type::state;
+		error = state_block_w_sideband.sideband.deserialize (stream);
+		assert (!error);
+
+		return state_block_w_sideband;
+	}
```
Comment (wezrule) [inline nano/secure/blockstore.hpp]: As this is essentially the same as `state_block_w_sideband_v14` it could be a refactored into a function template and then used in both places.

File: nano/secure/blockstore.cpp [hunk 4]
```diff
@@ -3,14 +3,54 @@
...
+	return result;
+}
+
+void nano::block_details::unpack (uint8_t details_a)
+{
+	is_send = static_cast<bool> (details_a >> 7);
+	is_receive = static_cast<bool> ((static_cast<uint8_t> (details_a << 1)) >> 7);
+	is_epoch = static_cast<bool> (static_cast<uint8_t> (details_a << 2) >> 7);
+	epoch = static_cast<nano::epoch> (static_cast<uint8_t> (details_a << 3) >> 3);
+}
```
Comment (wezrule) [inline nano/secure/blockstore.cpp]: Can using a `std::bitset<8>` instead of a `uint8_t` make this code a bit simpler? e.g can do:
```
is_send = bits.test (7);
is_receive = bits.test (6);
// etc
```

https://en.cppreference.com/w/cpp/utility/bitset

File: nano/secure/blockstore.cpp (hunk 1 above)
Comment (guilhermelawless) [reply nano/secure/blockstore.cpp]: Oops, good catch. I went with the first approach as we test block_details::deserialize independently.

File: nano/secure/blockstore.cpp (hunk 4 above)
Comment (guilhermelawless) [reply nano/secure/blockstore.cpp]: Thanks! Means converting to bitset and then to ulong but makes it less prone to errors (which I did initially).

File: nano/secure/ledger.cpp [hunk 5]
```diff
@@ -288,6 +289,7 @@ void ledger_processor::state_block_impl (nano::state_block const & block_a)
 						if (result.code == nano::process_result::progress)
 						{
 							is_send = block_a.hashables.balance < info.balance;
+							is_receive = block_a.hashables.balance > info.balance;
```
Comment (SergiySW) [inline nano/secure/ledger.cpp]: Receive block can be 0 balance change (if receiving old type block with 0 send)

File: nano/node/lmdb/lmdb.cpp [hunk 6]
```diff
@@ -781,6 +784,70 @@ void nano::mdb_store::upgrade_v16_to_v17 (nano::write_transaction const & transa
...
+		}
+		if (balance == prev_balance && network_params.ledger.epochs.is_epoch_link (block->link ()))
+		{
+			is_epoch = true;
+		}
+		else if (balance < prev_balance)
+		{
+			is_send = true;
+		}
+		else if (balance > prev_balance)
```
Comment (SergiySW) [inline nano/node/lmdb/lmdb.cpp]: Same here (receive)

File: nano/secure/ledger.cpp (hunk 5 above)
Comment (guilhermelawless) [reply nano/secure/ledger.cpp]: Is this check enough?
`is_receive = block_a.hashables.balance.number () >= info.balance.number () && !block_a.hashables.link.is_zero ();`

Comment (guilhermelawless) [reply nano/secure/ledger.cpp]: Added commit that also tests this case in the upgrade test

Comment (SergiySW) [reply nano/secure/ledger.cpp]: Seems enough

Comment (guilhermelawless) [reply nano/secure/ledger.cpp]: More specifically I did `is_receive = !is_send && !block_a.hashables.link.is_zero ();`

Should we add a `ledger` test for this edge case?

File: nano/nano_node/entry.cpp [hunk 7]
```diff
@@ -1050,6 +1050,44 @@ int main (int argc, char * const * argv)
...
+							}
+							else if (block->balance () == prev_balance && node.node->ledger.is_epoch_link (block->link ()))
+							{
+								// State epoch
+								block_details_error = !sideband.details.is_epoch || sideband.details.is_send || sideband.details.is_receive;
+							}
+							else
+							{
+								// State receive
+								block_details_error = !sideband.details.is_receive || sideband.details.is_send || sideband.details.is_epoch;
```
Comment (SergiySW) [inline nano/nano_node/entry.cpp]: I think we can add extra check ` !node.node->store.source_exists (transaction, block->link ()) `

Comment (guilhermelawless) [reply nano/nano_node/entry.cpp]: Updated

**Comment by guilhermelawless** [comment]
Changed recently so that open blocks are also marked as is_receive, for consistency.

@SergiySW performed a full validation of the blocks after bootstrapping from scratch on main net (thank you!).

**Comment by guilhermelawless** [comment]
Opening another PR as Github is being weird no matter what I do, and all commits are signed.
image
